### PR TITLE
Faster implementation of getValue

### PR DESCRIPTION
**Pull request by joehuchette**
cc @mlubin

### CODE REVIEW

**Comment by mlubin** [comment]
Does this replace #487?

**Comment by joehuchette** [comment]
aoz8kgx8pzknypz7z38n

File: src/JuMPDict.jl [hunk 1]
```diff
@@ -81,6 +81,8 @@ macro gendict(instancename,T,idxpairs,idxsets...)
         setidxrhs = :(setindex!(d.innerArray,val))
         maplhs = :(Base.map(f::Function,d::$(typename)))
         maprhs = :($(typename)(map(f,d.innerArray),d.name,d.indexsets,d.indexexprs))
+        _wraplhs = :(_wrapper(d::$(typename),inner)) # helper function that wraps array into JuMPArray of similar type
```
Comment (mlubin) [inline src/JuMPDict.jl]: How is this different from the definition below? Does it need to be defined here?

Comment (mlubin) [inline src/JuMPDict.jl]: Or can you define it for JuMPArrays below without the expression mashing?

Comment (joehuchette) [inline src/JuMPDict.jl]: I couldn't figure it out without triangular dispatch. I want something like
```jl
_wrapper{Typ<:JuMPArray,S}(d::Typ{S},inner) = Typ{Float64}(...)
```

**Comment by mlubin** [comment]
Could you rebase?

**Comment by joehuchette** [comment]
done

**Comment by mlubin** [comment]
Failed on 0.3:
```
ERROR: `eachindex` has no method matching eachindex(::Dict{(Any,Any,Any),Variable})
 in _getValueInner at /home/travis/.julia/v0.3/JuMP/src/JuMPDict.jl:158
 in anonymous at /home/travis/.julia/v0.3/JuMP/test/print.jl:177
 in facts at /home/travis/.julia/v0.3/FactCheck/src/FactCheck.jl:315
```

**Comment by joehuchette** [comment]
https://github.com/JuliaLang/Compat.jl/pull/118

File: src/JuMPDict.jl [hunk 2]
```diff
@@ -85,6 +85,8 @@ macro gendict(instancename,T,idxpairs,idxsets...)
         else
             maprhs = :($(typename)(map(f,d.innerArray),d.name,d.indexsets,d.indexexprs))
         end
+        _wraplhs = :(_wrapper(d::$(typename),inner)) # helper function that wraps array into JuMPArray of similar type
```
Comment (mlubin) [inline src/JuMPDict.jl]: Is there a better name we can use than ``_wrapper``? This is basically a constructor.

Comment (joehuchette) [inline src/JuMPDict.jl]: I'm open to suggestions, but it has to make sense for ``JuMPDict`` and ``OneIndexedArray`` as well.

Comment (mlubin) [inline src/JuMPDict.jl]: Well on 0.4, I think we can just use ``JuMPContainer``. Maybe ``JuMPContainer_from`` for now?

**Comment by joehuchette** [comment]
```jl
using JuMP

function bench(n)
    m = Model()
    @defVar(m, x[1:n,1:n,1:n], start=0)
    getValue(x)
end
```
On master:
```
julia> @time bench(100);
   1.510 seconds      (5000 k allocations: 141 MB, 51.07% gc time)

julia> @time bench(200);
  13.242 seconds      (40000 k allocations: 1124 MB, 47.71% gc time)
```

On this branch:
```
julia> @time bench(100);
 345.100 milliseconds (2000 k allocations: 97323 KB, 59.41% gc time)

julia> @time bench(200);
   5.392 seconds      (16000 k allocations: 878 MB, 40.06% gc time)
```

**Comment by mlubin** [comment]
Nice improvement. Why is there still so much GC churn?

**Comment by joehuchette** [comment]
No idea. Isn't there some way to track allocations now?

**Comment by IainNZ** [comment]
Maybe related to https://github.com/JuliaOpt/JuMP.jl/issues/437

**Comment by IainNZ** [comment]
Let me try tracking allocations

**Comment by IainNZ** [comment]
```
        - function Variable(m::Model,lower::Number,upper::Number,cat::Symbol,name::String="",value::Number=NaN)
        0     m.numCols += 1
 68157920     push!(m.colNames, name)
 68157920     push!(m.colNamesIJulia, name)
 68157920     push!(m.colLower, convert(Float64,lower))
 68157920     push!(m.colUpper, convert(Float64,upper))
 68157920     push!(m.colCat, cat)
196157920     push!(m.colVal,value)
        0     if cat == :Fixed
        0         @assert lower == upper
        0         m.colVal[end] = lower
        -     end
        0     if m.internalModelLoaded
        0         if method_exists(MathProgBase.addvar!, (typeof(m.internalModel),Vector{Int},Vector{Float64},Float64,Float64,Float64))
        0             MathProgBase.addvar!(m.internalModel,float(lower),float(upper),0.0)
        -         else
        0             Base.warn_once("Solver does not appear to support adding variables to an existing model. Hot-start is disabled.")
        0             m.internalModelLoaded = false
        -         end
        -     end
256000000     return Variable(m, m.numCols)
        - end
```

**Comment by IainNZ** [comment]
```
        - function _getValueInner(x)
        0     vars = _innercontainer(x)
 64000096     vals = _similar(vars)
        0     warnedyet = false
        0     for I in eachindex(vars)
        0         tmp = _getValue(vars[I])
        0         if isnan(tmp) && !warnedyet
        0             warn("Variable value not defined for entry of $(x.name). Check that the model was properly solved.")
        0             warnedyet = true
        -         end
        0         vals[I] = tmp
        -     end
        0     vals
        - end
```

**Comment by IainNZ** [comment]
```
        - function _local_index(indexsets, dim, k)
  3232999     n = length(indexsets)
        0     cprod = 1
        0     for i in 1:(dim-1)
        0         cprod *= length(indexsets[i])
        -     end
        0     idx = Compat.ceil(Integer, mod1(k,cprod*length(indexsets[dim])) / cprod)
        0     return indexsets[dim][idx]
        - end
        - 
        - function Base.next(x::JuMPArray,k)
        0     (var, gidx) = next(x.innerArray, k)
        0     key = map(i->_local_index(x.indexsets, i, k), 1:length(x.indexsets))
   341390     return (tuple(key..., var), gidx)
        - end
```

**Comment by IainNZ** [comment]
```
        - macro defVar(args...)
 10595819     length(args) <= 1 &&
        -         error("in @defVar: expected model as first argument, then variable information.")
        0     m = args[1]
        0     x = args[2]
   346022     extra = vcat(args[3:end]...)
      112     m = esc(m)
        - 
        0     t = :Cont
        0     gottype = 0
        -     # Identify the variable bounds. Five (legal) possibilities are "x >= lb",
        -     # "x <= ub", "lb <= x <= ub", "x == val", or just plain "x"
        0     if isexpr(x,:comparison)
        -         # We have some bounds
        0         if x.args[2] == :>= || x.args[2] == :≥
        0             if length(x.args) == 5
        -                 # ub >= x >= lb
        0                 x.args[4] == :>= || x.args[4] == :≥ || error("Invalid variable bounds")
        0                 var = x.args[3]
        0                 lb = esc_nonconstant(x.args[5])
        0                 ub = esc_nonconstant(x.args[1])
        -             else
        -                 # x >= lb
        0                 var = x.args[1]
        0                 @assert length(x.args) == 3
        0                 lb = esc_nonconstant(x.args[3])
        0                 ub = Inf
        -             end
        0         elseif x.args[2] == :<= || x.args[2] == :≤
        0             if length(x.args) == 5
        -                 # lb <= x <= u
        0                 var = x.args[3]
        0                 (x.args[4] != :<= && x.args[4] != :≤) &&
        -                     error("in @defVar ($var): expected <= operator after variable name.")
        0                 lb = esc_nonconstant(x.args[1])
        0                 ub = esc_nonconstant(x.args[5])
        -             else
        -                 # x <= ub
        0                 var = x.args[1]
        -                 # NB: May also be lb <= x, which we do not support
        -                 #     We handle this later in the macro
        0                 @assert length(x.args) == 3
        0                 ub = esc_nonconstant(x.args[3])
        0                 lb = -Inf
        -             end
        0         elseif x.args[2] == :(==)
        -             # fixed variable
        0             var = x.args[1]
        0             @assert length(x.args) == 3
        0             lb = esc(x.args[3])
        0             ub = esc(x.args[3])
        0             gottype = 1
        0             t = :Fixed
        -         else
        -             # Its a comparsion, but not using <= ... <=
     2240             error("in @defVar ($(string(x))): use the form lb <= ... <= ub.")
        -         end
        -     else
        -         # No bounds provided - free variable
        -         # If it isn't, e.g. something odd like f(x), we'll handle later
        0         var = x
        0         lb = -Inf
        0         ub = Inf
        -     end
        - 
        -     # separate out keyword arguments
   207333     kwargs = filter(ex->isexpr(ex,:kw), extra)
     1602     extra = filter(ex->!isexpr(ex,:kw), extra)
        - 
        -     # process keyword arguments
        0     value = NaN
        0     obj = nothing
        0     inconstraints = nothing
        0     coefficients = nothing
       32     for ex in kwargs
        0         if ex.args[1] == :start
      112             value = esc(ex.args[2])
        0         elseif ex.args[1] == :objective
        0             obj = esc(ex.args[2])
        0         elseif ex.args[1] == :inconstraints
        0             inconstraints = esc(ex.args[2])
        0         elseif ex.args[1] == :coefficients
        0             coefficients = esc(ex.args[2])
        -         else
        0             error("in @defVar ($var): Unrecognized keyword argument $(ex.args[1])")
        -         end
        -     end
        - 
        0     if (obj != nothing || inconstraints != nothing || coefficients != nothing) &&
        -         (obj == nothing || inconstraints == nothing || coefficients == nothing)
        0         error("in @defVar ($var): Must provide 'objective', 'inconstraints', and 'coefficients' arguments all together for column-wise modeling")
        -     end
        - 
   652643     sdp = any(t -> (t == :SDP), extra)
```

**Comment by IainNZ** [comment]
those are all the non-negligible allocations I spotted

**Comment by mlubin** [comment]
Does ``@code_warntype`` report anything wrong?

**Comment by IainNZ** [comment]
Where would you stick the `@code_warntype`?
`@code_warntype(@defVar(m, x[1:n,1:n,1:n], start=0))` gives me
```
ERROR: LoadError: UndefVarError: x not defined
 in bench at /Users/idunning/so.jl:5
 in include at ./boot.jl:254
 in include_from_node1 at ./loading.jl:197
 in process_options at ./client.jl:308
 in _start at ./client.jl:411
while loading /Users/idunning/so.jl, in expression starting on line 172
```

**Comment by mlubin** [comment]
_local_index is doing *a lot* of work

**Comment by mlubin** [comment]
On the getValue

**Comment by IainNZ** [comment]
```
256000000     return Variable(m, m.numCols)
```
Thats 256MB, right? Thats wild

**Comment by joehuchette** [comment]
Are the line numbers off? I can't imagine ``length(args) <= 1 && error("in @defVar: expected model as first argument, then variable information.")`` triggers so many allocations.

``_local_index`` is a really kludgy way of doing iteration. I've been thinking about a better way to do it but haven't really hit on anything. Some of the new array indexing stuff will probably help.

**Comment by joehuchette** [comment]
Although why is this benchmark even doing iteration over a JuMPContainer...?

**Pull request merged**